Add rosconsole command to change logger levels from command line

cmansley opened the pull request:
There is only the rqt_logger_level command for changing the logger
level for node. If you are doing remote work without an X session, you
have to send service commands directly or change the logger
configuration. This command line tool allows you to query and set the
logger level.

Commit:
Add rosconsole command to change logger levels from command line

There is only the rqt_logger_level command for changing the logger
level for node. If you are doing remote work without an X session, you
have to send service commands directly or change the logger
configuration. This command line tool allows you to query and set the
logger level.

ros-pull-request-builder commented:
Can one of the admins verify this patch?

Commit:
Add installation for rosconsole to CMakeLists

dirk-thomas commented on tools/rosconsole/CMakeLists.txt:
```diff
@@ -121,3 +121,8 @@ if(CATKIN_ENABLE_TESTING)
   endif()
 endif()
+
+catkin_python_setup()
```
The package needs a `setup.py` file for this to work. You might have missed to add it to the repo.

dirk-thomas commented on tools/rosconsole/CMakeLists.txt:
```diff
@@ -121,3 +121,8 @@ if(CATKIN_ENABLE_TESTING)
+catkin_python_setup()
+
+catkin_install_python(PROGRAMS scripts/rosconsole
+  DESTINATION ${CATKIN_PACKAGE_BIN_DESTINATION})
```
It probably makes sense to install this into the global bin folder rather then the package-specific libexec folder to be able to run it without `rosrun`. Instead of using `catkin_install_python` you can list it as a script in the setup.py file.

dirk-thomas commented on tools/rosconsole/src/rosconsole/__init__.py:
```diff
@@ -0,0 +1,144 @@
+#!/usr/bin/env python
+
+# make sure we aren't using floor division
+from __future__ import division, print_function
```
It looks like you need neither of these two.

dirk-thomas commented on tools/rosconsole/src/rosconsole/__init__.py:
```diff
@@ -0,0 +1,144 @@
+    if argv is None:
+        argv=sys.argv
+    # filter out remapping arguments in case we are being invoked via roslaunch
+    argv = rospy.myargv(argv)
```
Since all nodes will "race" when being invoked through `roslaunch` it might not be a good idea to use this script to try to change the logger level since the targeted node might not have started fast enough.

cmansley replied:
Do you have a suggestion to fix this in the tool? Because it would be a shame to lose out on the functionality of the tool just because of the possibility of abuse.

dirk-thomas replied:
You can keep this line in the patch. I just wanted to state that using it in a launch file is imo not recommended. Therefore the comment is a bit misleading.

roslaunch is not designed to handle this and will very likely not be modified to do so. We are aware of that limitation and ROS 2 will for sure address those but that won't help in ROS 1.

cmansley replied:
Another copy and paste artifact. I can remove the comment.

dirk-thomas commented:
This is a great new feature.

I made a couple of small comments inline. The "biggest" suggestion would be to use `argparse` instead of custom argv handling. This will likely also automatically improve the usage information for users.

Commit:
Remove unnecessary imports

Commit:
Add setup.py file

Commit:
Remove installing to the package specific folder

cmansley commented:
I fixed some of the copy and paste errors and I added the setup.py file. Thanks for the feedback. 

With regards to argparse, I simply copied what exists in the other default ros command line tools. And at least on older systems, argparse is not part of the default python install.

dirk-thomas commented:
Actually nowadays it is part of it. Since this patch targets Indigo which implies Ubuntu Trusty which comes with Python 2.7 which includes argparse by default.

Anyway you don't have to use `argparse` - it was only meant as a suggestion. A lot of existing tools still use `optparse`.

Commit:
Remove misleading comment that was copied from another tool

dirk-thomas commented on tools/rosconsole/setup.py:
```diff
@@ -0,0 +1,13 @@
+    packages=['rosconsole'],
+    package_dir={'': 'src'},
+    scripts=['scripts/rosconsole'],
+    requires=['genmsg', 'genpy', 'roslib', 'rospkg']
```
Can you please explain why you used these four dependencies?

cmansley replied:
I copied from all other default tools.

dirk-thomas commented:
Please make new Python code comply  with PEP8.

Commit:
pep8 the python scripts

Note: pep8 does allow an "increase the nominal line length from 80 to
100 characters"

gerkey commented:
Thanks for the contribution.  I made a few small fixes and created #594 to replace this PR.

ros-pull-request-builder commented:
Can one of the admins verify this patch?
